app/supplier_utils.py

```diff
@@ -37,3 +44,80 @@ def check_supplier_role(role, supplier_id):
         abort(400, "'supplierId' is required for users with 'supplier' role")
     elif role != 'supplier' and supplier_id:
         abort(400, "'supplierId' is only valid for users with 'supplier' role, not '{}'".format(role))
+
+
+def company_details_confirmed_if_required_for_framework(framework_slug, supplier_framework):
+    return (
+        supplier_framework.supplier.company_details_confirmed
+        if framework_slug == 'g-cloud-10' else
+        supplier_framework.application_company_details_confirmed
+    )
+
+
+def update_open_declarations_with_company_details(db, supplier_id, updater_json, specific_framework_slug=None):
+    """Expected to be called within a view"""
+    open_supplier_frameworks_query = SupplierFramework.query.filter(
+        SupplierFramework.supplier_id == supplier_id,
+    )
+
+    if specific_framework_slug:
+        open_supplier_frameworks_query = open_supplier_frameworks_query.filter(
+            SupplierFramework.framework.has(Framework.slug == specific_framework_slug),
+        )
+    else:
+        open_supplier_frameworks_query = open_supplier_frameworks_query.filter(
+            SupplierFramework.framework.has(Framework.status == 'open'),
+        )
+
+    # We need to ensure that SupplierFramework and Framework haven't changed during this transaction, so grab
+    # the affected rows with a for update lock.
+    open_supplier_frameworks = open_supplier_frameworks_query.options(lazyload('*')).with_for_update().all()
+
+    if not open_supplier_frameworks:
+        return
+
+    supplier = Supplier.query.filter(
+        Supplier.supplier_id == supplier_id
+    ).first_or_404()
+
+    # Update the open framework application(s) with latest company details.
+    for open_supplier_framework in open_supplier_frameworks:
+        company_details = {
+            supplier_constants.KEY_TRADING_NAME: supplier.name,
+            supplier_constants.KEY_REGISTERED_NAME: supplier.registered_name,
+            supplier_constants.KEY_REGISTRATION_NUMBER: (
+                supplier.companies_house_number or supplier.other_company_registration_number
+            ),
+            supplier_constants.KEY_DUNS_NUMBER: supplier.duns_number,
+            supplier_constants.KEY_VAT_NUMBER: supplier.vat_number,
+            supplier_constants.KEY_TRADING_STATUS: supplier.trading_status,
+            supplier_constants.KEY_ORGANISATION_SIZE: supplier.organisation_size,
+            supplier_constants.KEY_REGISTRATION_COUNTRY: supplier.registration_country,
+            supplier_constants.KEY_REGISTRATION_BUILDING: supplier.contact_information[0].address1,
+            supplier_constants.KEY_REGISTRATION_TOWN: supplier.contact_information[0].city,
+            supplier_constants.KEY_REGISTRATION_POSTCODE: supplier.contact_information[0].postcode,
+        }
+
+        company_details_without_nones = {k: v for k, v in company_details.items() if v is not None}
+
+        open_supplier_framework.declaration = {
```

Comment:
Why not `open_supplier_framework.declaration.update(company_details_without_nones)`?

Reply:
Also, hypothetical situation, but if a supplier initially put a `companies_house_number`, and that got added to the declaration, then they realised that it should have been a `other_company_registration_number`, or visa versa, it there a risk of the old key being left in the declaration?

Reply:
`open_supplier_framework.declaration.update(company_details_without_nones)` - this encounters the issue with Postgres/SQLAlchemy not recognising changes within JSON blobs. Creating a new dict altogether means that the changes are recognised and persisted to the db correctly.

Reply:
Cool. But regarding my point about companies house numbers, if the declaration had already been updated with a companies house number, is the unpacking method going to remove the old key? I don't think either way would...

Reply:
Sorry, forgot to address that bit.

When a supplier updates either of these numbers, the other one is wiped out from the Supplier record. This method then denormalizes the two fields (specifically, taking the value of whichever is set) under a single name 'supplierCompanyRegistrationNumber', which can be _either_ a companies house number or 'another' company registration number.

So no, there's no chance of both of them ending up here in the declaration at the same time.